File: backend/app/api/endpoints/ai.py
"""
ai.py — Phase 7: /api/ai/edit endpoint.

POST /api/ai/edit
  Body (JSON):
    {
      "prompt":    str   required  — editing instruction e.g. "make the sky dramatic"
      "image_uri": str   required  — data-URI of the photo (data:image/jpeg;base64,...)
    }

  Returns:
    {
      "url":               str   — CDN URL of the AI-edited image (~1 hour expiry)
      "credits_remaining": int
    }

Rate limiting: AI_RATE_LIMIT_PER_HOUR requests per user per hour (default 10).
In DEV_BYPASS_AUTH mode user_id is "dev_user".
"""
from __future__ import annotations

from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel, field_validator

from app.core.config import settings
from app.services.ai_service import (
    check_rate_limit,
    decode_image_uri,
    edit_image,
    get_credits_remaining,
)

router = APIRouter()


# ---------------------------------------------------------------------------
# Auth helper
# ---------------------------------------------------------------------------

def _resolve_user_id(authorization: str | None) -> str:
    if settings.DEV_BYPASS_AUTH:
        return "dev_user"

    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="Missing or invalid Authorization header")

    token = authorization.removeprefix("Bearer ").strip()

    try:
        import jwt  # type: ignore[import]

        if not settings.JWT_SECRET:
            raise HTTPException(status_code=500, detail="JWT_SECRET is not configured")

        payload = jwt.decode(token, settings.JWT_SECRET, algorithms=["HS256"])
        user_id: str = payload.get("sub") or payload.get("user_id") or ""
        if not user_id:
            raise HTTPException(status_code=401, detail="Token has no user_id/sub claim")
        return user_id
    except Exception as exc:
        raise HTTPException(status_code=401, detail=f"Invalid token: {exc}") from exc


# ---------------------------------------------------------------------------
# Schemas
# ---------------------------------------------------------------------------

class EditRequest(BaseModel):
    prompt: str
    image_uri: str  # required for edit

    @field_validator("prompt")
    @classmethod
    def prompt_not_empty(cls, v: str) -> str:
        cleaned = v.strip()
        if not cleaned:
            raise ValueError("prompt must not be empty")
        if len(cleaned) > 1000:
            raise ValueError("prompt must be 1000 characters or fewer")
        return cleaned

    @field_validator("image_uri")
    @classmethod
    def image_uri_not_empty(cls, v: str) -> str:
        if not v or not v.startswith("data:"):
            raise ValueError("image_uri must be a base64 data-URI")
        return v


class EditResponse(BaseModel):
    b64: str           # data-URI of the edited image (data:image/png;base64,...)
    credits_remaining: int


# ---------------------------------------------------------------------------
# Endpoint
# ---------------------------------------------------------------------------

@router.post("/ai/edit", response_model=EditResponse)
async def ai_edit(
    body: EditRequest,
    authorization: str | None = Header(default=None),
) -> EditResponse:
    """
    Edit a photo with AI using OpenAI gpt-image-1.

    The client sends the current photo as a base64 data-URI together with
    a natural-language editing instruction. The backend forwards the image
    and prompt to OpenAI, then returns the generated image as a data URI.
    """
    if not settings.OPENAI_API_KEY:
        raise HTTPException(
            status_code=503,
            detail="AI service is not configured. Set OPENAI_API_KEY in server .env.",
        )

    user_id = _resolve_user_id(authorization)

    # Rate-limit check
    allowed, _ = check_rate_limit(user_id)
    if not allowed:
        raise HTTPException(
            status_code=429,
            detail={
                "error": "rate_limit_exceeded",
                "message": f"You have used all {settings.AI_RATE_LIMIT_PER_HOUR} AI edits for this hour. Please try again later.",
                "credits_remaining": 0,
            },
        )

    credits_remaining = get_credits_remaining(user_id)

    try:
        image_b64 = decode_image_uri(body.image_uri)
        result = edit_image(body.prompt, image_b64)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc
    except Exception as exc:
        error_msg = str(exc)
        if "quota" in error_msg.lower() or "billing" in error_msg.lower() or "insufficient_quota" in error_msg.lower():
            raise HTTPException(
                status_code=402,
                detail="OpenAI quota exceeded. Please check your billing settings.",
            ) from exc
        if "content_policy" in error_msg or "safety" in error_msg.lower() or "prohibited" in error_msg.lower():
            raise HTTPException(
                status_code=400,
                detail="Your prompt was rejected by the content safety filter. Try a different instruction.",
            ) from exc
        raise HTTPException(
            status_code=500,
            detail=f"AI edit failed: {error_msg}",
        ) from exc

    return EditResponse(
        b64=result["b64"],
        credits_remaining=credits_remaining,
    )

File: src/lib/api/ai.ts
/**
 * ai.ts — Phase 7: Mobile client for /api/ai/edit.
 *
 * Usage:
 *   import { editPhotoWithAI } from '@/src/lib/api/ai';
 *   const { url } = await editPhotoWithAI({ prompt: 'Make the sky dramatic', imageUri: 'file://...' });
 */

const BACKEND_URL = process.env.EXPO_PUBLIC_BACKEND_URL?.trim();

export interface AIEditRequest {
  /** Natural-language editing instruction */
  prompt: string;
  /** Base64 data-URI of the photo (data:image/jpeg;base64,...) */
  imageUri: string;
}

export interface AIEditResult {
  /** Base64 data-URI of the AI-edited image (data:image/png;base64,...) */
  b64: string;
  creditsRemaining: number;
}

export class AIError extends Error {
  constructor(
    message: string,
    public readonly code:
      | 'not_configured'
      | 'rate_limited'
      | 'content_policy'
      | 'quota_exceeded'
      | 'unauthorized'
      | 'network_error'
      | 'unknown',
    public readonly creditsRemaining?: number,
  ) {
    super(message);
    this.name = 'AIError';
  }
}

export async function editPhotoWithAI(
  request: AIEditRequest,
  authToken?: string,
): Promise<AIEditResult> {
  if (!BACKEND_URL) {
    throw new AIError(
      'Backend is not configured. Set EXPO_PUBLIC_BACKEND_URL in your .env.',
      'not_configured',
    );
  }

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };

  if (authToken) {
    headers['Authorization'] = `Bearer ${authToken}`;
  }

  let response: Response;

  try {
    response = await fetch(`${BACKEND_URL}/api/ai/edit`, {
      method: 'POST',
      headers,
      body: JSON.stringify({
        prompt: request.prompt,
        image_uri: request.imageUri,
      }),
    });
  } catch {
    throw new AIError(
      'Could not reach the backend. Make sure the server is running.',
      'network_error',
    );
  }

  if (response.ok) {
    const data = await response.json();
    return {
      b64: data.b64,
      creditsRemaining: data.credits_remaining ?? 0,
    };
  }

  let errorData: Record<string, unknown> = {};
  try { errorData = await response.json(); } catch { /* ignore */ }

  const detail = errorData['detail'];

  if (response.status === 401) {
    throw new AIError('You must be logged in to use AI editing.', 'unauthorized');
  }
  if (response.status === 402) {
    throw new AIError('Gemini quota exceeded. Please check the server billing settings.', 'quota_exceeded');
  }
  if (response.status === 429) {
    const inner = typeof detail === 'object' && detail !== null ? (detail as Record<string, unknown>) : {};
    const remaining = typeof inner['credits_remaining'] === 'number' ? inner['credits_remaining'] : 0;
    throw new AIError(
      typeof inner['message'] === 'string' ? inner['message'] : 'You have used all your AI edits for this hour.',
      'rate_limited',
      remaining,
    );
  }
  if (response.status === 400) {
    const msg = typeof detail === 'string' ? detail : 'Your prompt was rejected by the content safety filter.';
    throw new AIError(msg, 'content_policy');
  }

  throw new AIError(
    typeof detail === 'string' ? detail : `Unexpected error (HTTP ${response.status})`,
    'unknown',
  );
}
